### PR TITLE
docs: SMI Proposal

### PR DESCRIPTION
**Pull request by dthomson25**
This PR introduces a proposal to support SMI and is the first part of solving https://github.com/argoproj/argo-rollouts/issues/359. 

At a high level, the proposal suggests add the following an SMI section to the trafficRouting section of the Rollout:
```yaml
  strategy:
    canary:
      steps:
      - setWeight: 5
      - pause:
          duration: 600
      canaryService: canary-svc # required
      stableService: stable-svc # required
      trafficRouting:
        # New fields added
        smi:
         rootService: root-svc # required
         trafficSplitName: rollout-example-traffic-split # optional
```

which then creates a TrafficSplit with the following yaml:
```yaml
apiVersion: split.smi-spec.io/v1alpha1
kind: TrafficSplit
metadata:
  name: rollout-example-traffic-split
spec:
  service: root-svc
  backends:
  - service: stable-svc
    weight: 5
  - service: canary-svc
    weight: 95
```

### CODE REVIEW

File: docs/features/traffic-management/smi.md [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    canary:
+      steps:
+      - setWeight: 5
+      - pause:
+          duration: 600
+      canaryService: canary-svc # required
+      stableService: stable-svc # required
+      trafficRouting:
+        smi:
+         rootService: root-svc # required
```
Comment (grampelberg) [inline docs/features/traffic-management/smi.md:33]: Any reason to expose the `root-svc` to users? You should be able to treat `stable-svc` the same way, and then this all works for clients that don't have traffic split support (eg. anything unmeshed).

File: docs/features/traffic-management/smi.md [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+spec:
+  service: root-svc
+  backends:
+  - service: stable-svc
+    weight: 5
+  - service: canary-svc
+    weight: 95
+```
+
+As a Rollout progresses through all its steps, the controller updates the TrafficSplit's backend weights to reflect the current weight of the Rollout. When the Rollout has successfully finished executing all the steps, the controller modifies the TrafficSplit to send 100% of traffic to the stable Service.
```
Comment (grampelberg) [inline docs/features/traffic-management/smi.md:55]: You might want to talk through what happens with the labels and how traffic is actually handled under the covers. This is a great overview but I find the picture super helpful for explaining how the system fits together.

File: docs/features/traffic-management/smi.md (hunk 1 above)
Comment (dthomson25) [reply docs/features/traffic-management/smi.md:33]: Are you suggesting something like the following?
```yaml
apiVersion: split.smi-spec.io/v1alpha1
kind: TrafficSplit
metadata:
  name: rollout-example-traffic-split
spec:
  service: stable-svc
  backends:
  - service: stable-svc
    weight: 95
  - service: canary-svc
    weight: 5
```
Wouldn't that violate the fourth tradeoff listed in the [SMI TrafficSplit spec](https://github.com/servicemeshinterface/smi-spec/blob/master/traffic-split.md#tradeoffs):
```
* TrafficSplits cannot be self-referential
```

If having a self-referential TrafficSplit is allowed, then the rootService could be optional and it should reference the stableService if anything.

File: docs/features/traffic-management/smi.md (hunk 2 above)
Comment (dthomson25) [reply docs/features/traffic-management/smi.md]: Great suggestion! I added a picture to make it a little more clear what components are involved. Let me know what you think!

File: docs/features/traffic-management/smi.md (hunk 1 above)
Comment (grampelberg) [reply docs/features/traffic-management/smi.md:33]: > * TrafficSplits cannot be self-referential

This is specifically related to recursion, you can't have a traffic split that forwards to a traffic split. In retrospect, I probably could have worded that a little bit better.

> If having a self-referential TrafficSplit is allowed, then the rootService could be optional and it should reference the stableService if anything.

I think you're right. I'll need to double check out Flagger's doing it, but I believe that's the way they're making it all work under the covers.

File: docs/features/traffic-management/smi.md (hunk 2 above)
Comment (grampelberg) [reply docs/features/traffic-management/smi.md]: Cool! That's super helpful!

File: docs/features/traffic-management/smi.md (hunk 1 above)
Comment (dthomson25) [reply docs/features/traffic-management/smi.md]: Great! Just pushed the change to make the rootService optional

**Comment by dthomson25** [comment]
@grampelberg, unless you have any other feedback, I'll merge the proposal in and @khhirani will do the implementation! Thanks again for taking the time to review!

**Comment by grampelberg** [comment]
@dthomson25 go for it, thank you for involving me in the process! I'd love to help out during implementation any way I can =0

**Pull request merged**